osc/osc_test.go

```diff
@@ -111,53 +113,80 @@ func TestAddMsgHandlerWithInvalidAddress(t *testing.T) {
 	}
 }
 
+// These tests stop the server by forcibly closing the connection, which causes
+// a "use of closed network connection" error the next time we try to read from
+// the connection. As a workaround, this wraps server.ListenAndServe() in an
+// error-handling layer that doesn't consider "use of closed network connection"
+// an error.
+//
+// Open question: is this desired behavior, or should server.serve return
+// successfully in cases where it would otherwise throw this error?
+func serveUntilInterrupted(server *Server) error {
+	if err := server.ListenAndServe(); err != nil &&
+		!strings.Contains(err.Error(), "use of closed network connection") {
+		return err
+	}
+
+	return nil
+}
+
 func TestServerMessageDispatching(t *testing.T) {
 	finish := make(chan bool)
 	start := make(chan bool)
 	done := sync.WaitGroup{}
 	done.Add(2)
 
-	// Start the OSC server in a new go-routine
-	go func() {
-		conn, err := net.ListenPacket("udp", "localhost:6677")
-		if err != nil {
-			t.Fatal(err)
-		}
-		defer conn.Close()
+	port := 6677
+	addr := "localhost:" + strconv.Itoa(port)
+
+	server := &Server{Addr: addr, Dispatcher: NewStandardDispatcher()}
+	defer server.CloseConnection()
+
+	if err := server.Dispatcher.(*StandardDispatcher).AddMsgHandler(
+		"/address/test",
+		func(msg *Message) {
+			defer func() {
+				server.CloseConnection()
+				finish <- true
+			}()
 
-		d := NewStandardDispatcher()
```

Comment:
Any reason for moving away from assigning `NewStandardDispatcher()` to `d` and then adding a handler?
I'm wondering if it may be more obvious to less-proficient members of the Go community whether it would be best to keep it as it previously way and avoid the type assertion.

I feel like with tests usually it's best to optimise for clarity.

Reply:
The trouble I ran into is that the message handler function needs to have a reference to `server`, so `server` needs to be defined first.

I just realized that I can solve both problems, though, by doing it this way:

```diff
diff --git a/osc/osc_test.go b/osc/osc_test.go
index a979eb1..1b19ed0 100644
--- a/osc/osc_test.go
+++ b/osc/osc_test.go
@@ -139,9 +139,10 @@ func TestServerMessageDispatching(t *testing.T) {
 	port := 6677
 	addr := "localhost:" + strconv.Itoa(port)
 
-	server := &Server{Addr: addr, Dispatcher: NewStandardDispatcher()}
+	d := NewStandardDispatcher()
+	server := &Server{Addr: addr, Dispatcher: d}
 
-	if err := server.Dispatcher.(*StandardDispatcher).AddMsgHandler(
+	if err := d.AddMsgHandler(
 		"/address/test",
 		func(msg *Message) {
 			defer func() {
```

I wasn't sure if that would work, but it totally does, because `server.Dispatcher` is just a pointer to `d`.

Reply:
Can we close the server after we send the message from the client anyway?

Reply:
No, because that would be subject to a race condition. It's possible that we would send the message from the client, and then close the server before the server has a chance to receive the message.

Reply:
I feel like that's what the WaitGroup should be protecting. We would ensure all of the logic for the test is completed and then close the server once we're happy or unhappy with the results. 

We could assign the server to a variable and then on the next line create a `defer` function that closes the server. Then whichever route out of the test function, we ensure we are always closing the server last.

Either way, this isn't something that would block this PR, from my perspective.

Reply:
I took a closer look at this just now and you're totally right! I've fixed it up so that we just do a `defer` function right after we define the server, that closes the server.